File: quicksort.py
import random

array = [4, 3, 5, 7, 6, 2, 1, 9, 8, 10]

def partition(array, left, right, p):
    pivot = array[p]
    array[p] = array[left]
    array[left] = pivot
    p = left
    left += 1

    while right != left:
        # Move left pointer until an element greater than pivot is found
        while array[left] <= pivot and right != left:
            left += 1

        # Move right pointer until an element less than or equal to pivot is found
        while array[right] > pivot and right != left:
            right -= 1

        # Swap the elements at left and right pointers
        holder = array[left]
        array[left] = array[right]
        array[right] = holder

    if array[left] > pivot:
        # Swap pivot with the element at the left pointer - 1 position
        holder = array[left - 1]
        array[left - 1] = pivot
        array[p] = holder
        return left - 1
    else:
        # Swap pivot with the element at the left pointer position
        holder = array[left]
        array[left] = pivot
        array[p] = holder
        return left

def quicksort(array, left, right):
    if right - left == 1:
        # Base case: If there are only two elements, swap them if necessary
        if array[left] > array[right]:
            holder = array[right]
            array[right] = array[left]
            array[left] = holder
        return

    elif right - left < 1:
        # Base case: If there are no elements or only one element, return
        return

    else:
        # Choose a random pivot index between left and right
        p = random.randrange(left, right)
        # Partition the array and get the updated pivot index
        p = partition(array, left, right, p)

        # Recursively sort the subarrays before and after the pivot
        quicksort(array, left, p - 1)
        quicksort(array, p + 1, right)

# Print the original array
print("Original array:", array)

# Sort the array using quicksort
quicksort(array, 0, len(array) - 1)

# Print the sorted array
print("Sorted array:", array)

File: Quicksort.java
package Sorting;
import java.util.Random;
public class Quicksort {
    /**
     * Quicksort algorithm:
     * - Selects a pivot element from the array.
     * - Partitions the array into two subarrays: elements smaller than the pivot and elements greater than the pivot.
     * - Recursively applies the Quicksort algorithm to the subarrays.
     * - Combines the sorted subarrays to obtain the final sorted array.
     */
    public static void quickSort(int[] array, int low, int high ){
        if(low<high){
            int pivot=partition(array, low, high);
            quickSort(array, low, pivot-1); //sorts the left side of the pivot of the array
            quickSort(array, pivot+1, high); //sorts the right side of the pivot of the array
        }
    }
    public static int partition(int [] array, int low, int high){
        int pivot = array[high]; //taking the pivot element, can be subsituted for a random number
        int i = low-1;  //index of smaller elements
        int z = low;    //index that goes through the smaller array
        while(z<high){
            //swaps the smaller element w an element left of the pivot
            if(array[z]<pivot){
            int temp=array[i+1];
            array[i+1]=array[z];
            array[z]=temp;
            i++;
            }   
            z++;
        }
        //swaps pivot with last element sorted to make sure it is in correct place
        int temp = array[i + 1];
        array[i + 1] = array[high];
        array[high] = temp;
        return i; //returns index of pivot
        
    }
    public static void main(String[]args){
        int[]tester={6,3,1,4,9,10,7,8,5,2};
        quickSort(tester, 0, tester.length-1);
        for(int i = 0; i<tester.length; i++){
            System.out.println(tester[i]);
        }
    }
}
